### ISSUE TITLE
performance regression with nested arrays

### ISSUE THREAD

**Issue by my-little-repository** [JuliaLang/julia#10356]
I just updated to julia 0.4.0-dev+3622 (2015-02-27 03:37 UTC) from 0.3.6. I see a huge performance regression when it comes to recursion with arrays, as in the following example, where a recursive array of depth 20 is built then flattened.

      julia> gen_tree(n) = n == 0 ? 0 : {{gen_tree(n-1)},{gen_tree(n-1)}}    # binary tree with 0 as leaves
      julia> flatten(A) =  [[isa(x,Array)? flatten(x): x for x in A]...]              # flatten such a tree
      julia> @time length(flatten(gen_tree(20)))
      elapsed time: 415.118361468 seconds (1285233 MB allocated, 12.80% gc time in 58262 pauses with 9 full sweep)
      1048576

I was around 1.6s with julia version 0.3.6. Replacing braces by square brackets in the definition of gen_tree gives slightly better performance here (65s) but it is still pretty bad. In v0.3, there was no gain replacing {} by [] as far as I can remember.

Side note: I know that there are better ways to flatten a recursive array. This is just a short illustration of the regression.

**Comment by timholy** [comment]
If you have any deprecation warnings, those cause a *huge* performance hit. http://docs.julialang.org/en/latest/manual/performance-tips/#fix-deprecation-warnings

**Comment by my-little-repository** [comment]
No, I don't have deprecation warnings. I should have though if I use the braces. Weird. Maybe the recursive nature of the algorithm flushes them. This may explain the difference between a gen_tree with braces and brackets. So let me report the benchmarks with the brackets.

        julia> gen_tree(n) = n == 0 ? 0 : [[gen_tree(n-1)] ; [gen_tree(n-1)]];
        julia> flatten(A) =  [[isa(x,Array)? flatten(x): x for x in A]...];
        julia> @time flatten(gen_tree(20));
        elapsed time: 170.75246294 seconds (642721 MB allocated, 25.59% gc time in 29133 pauses with 1 full sweep)

**Comment by simonster** [comment]
With a fresh launch, I see deprecation warnings for the concatenation (although oddly not for the braces):

```julia
julia> gen_tree(n) = n == 0 ? 0 : {{gen_tree(n-1)},{gen_tree(n-1)}};

julia> flatten(A) =  [[isa(x,Array)? flatten(x): x for x in A]...];

julia> @time length(flatten(gen_tree(20)))
WARNING: [a,b] concatenation is deprecated; use [a;b] instead
 in depwarn at ./deprecated.jl:40
 in oldstyle_vcat_warning at ./abstractarray.jl:26
 in vect at abstractarray.jl:29
 in flatten at none:1 (repeats 39 times)
WARNING: [a] concatenation is deprecated; use [a;] instead
 in depwarn at ./deprecated.jl:40
 in oldstyle_vcat_warning at ./abstractarray.jl:26
 in vect at abstractarray.jl:29
 in flatten at none:1 (repeats 38 times)
```

(I cancelled before it finished.) Fixing the concatenation by adding a semicolon appears to make things faster:

```julia
julia> flatten(A) =  [[isa(x,Array)? flatten(x): x for x in A]...;];

julia> @time length(flatten(gen_tree(20)))
elapsed time: 2.323453212 seconds (867 MB allocated, 22.05% gc time in 34 pauses with 4 full sweep)
1048576
```

**Comment by my-little-repository** [comment]
@simonster Ah, perfect, that's the missing semicolon (or the warning generated) that is responsible for the slowdown. Thanks. Is there a way to shutdown the warnings, so as to prevent the slowdown?

On a slightly different topic, I must say that with the removal of the braces, nested arrays (think json) have become pretty hard to read, e.g.

       Any[Any[Any[0,0],Any[0,0]],Any[Any[0,0],Any[0,0]]]

instead of

       {{{0,0},{0,0}},{{0,0},{0,0}}}

**Comment by timholy** [comment]
Try `--depwarn=no`. But keep in mind that warnings today will eventually turn into errors, without the helpful tips about how you should change your code. Might as well make the changes now. In places where the new syntax is not available in julia 0.3, check the Compat package.
